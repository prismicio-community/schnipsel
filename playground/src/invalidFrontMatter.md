---
name: "Hello World Invalid"
description: "Says hello to the world invalid"
scopes: false
prefix: 42
---

```typescript
export const hello = (str: string): void => {
	console.info("Hello %s", str);
};
```
