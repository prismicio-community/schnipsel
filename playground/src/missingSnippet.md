---
name: "Hello World Missing"
description: "Says hello to the world missing"
scope: ["typescript"]
prefix: "helloWorldMissing"
---
