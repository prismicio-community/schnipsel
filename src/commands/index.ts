export { _default } from "./default";
export { init } from "./init";
export { help } from "./help";
