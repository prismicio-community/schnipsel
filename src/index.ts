export { run, runWatch } from "./process";

export { ConfigFile } from "./ConfigFile";
export { Input } from "./Input";

export type { Config, SnippetObject } from "./types";
