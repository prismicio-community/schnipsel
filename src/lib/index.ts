export { JSONFile, JSONFileWithMeta } from "./JSONFile";
export { lineBreak } from "./lineBreak";
export { logger } from "./logger";
export { debug } from "./debug";
