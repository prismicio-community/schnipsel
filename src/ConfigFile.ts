import chalk from "chalk";
import { loadConfig } from "c12";

import { JSONFile, debug, AnyFile } from "./lib";
import { CONFIG_FILE, NAME } from "./const";
import { SchnipselConfig } from "./types";
import { Renderers } from "./renderers";

export class ConfigFile extends AnyFile<SchnipselConfig> {
	constructor(cwd = process.cwd()) {
		super(CONFIG_FILE, cwd);
	}

	async init(): Promise<void> {
		if (this.exists()) {
			throw new Error(
				`${chalk.cyan(this.name)} already exists in ${chalk.cyan(this.cwd)}`,
			);
		}

		const { indent } = await new JSONFile("package.json", this.cwd).readMeta();

		await this._write(
			`import { defineSchnipselConfig } from "schnipsel";

export default defineSchnipselConfig({
	input: {
		directory: "src",
	},

	// Update or remove renderers according to desired result...
	renderers: [
		{
			name: "vscode",
			options: {
				outputDirectory: "vscode/snippets",
				packageJSON: "vscode/package.json",
			},
		},
		{
			name: "sublime",
			options: {
				outputDirectory: "sublime/snippets",
			},
		},
		{
			name: "intellij",
			options: {
				outputDirectory: "intellij/src/main/resources/liveTemplates",
				pluginXML: "intellij/src/main/resources/META-INF/plugin.xml",
			},
		},
	],
});
`.replace(/\t/g, indent),
		);
	}

	async read(): Promise<SchnipselConfig> {
		debug("Reading config file...");

		if (!this.exists()) {
			throw new Error(
				`${chalk.cyan(this.name)} not found in ${chalk.cyan(
					this.cwd,
				)}, create one with ${chalk.cyan(`${NAME} init`)}`,
			);
		}

		const { config } = await loadConfig({
			name: "schnipsel",
			configFile: "schnipsel.config",
			rcFile: false,
			globalRc: false,
			cwd: this.cwd,
		});

		this.validate(config);

		debug("Config file read and validated!");

		return config;
	}

	async write(_config: SchnipselConfig): Promise<void> {
		throw new Error("`ConfigFile.write()` is not supported");
	}

	validate(
		maybeConfig: Record<string, unknown>,
	): asserts maybeConfig is SchnipselConfig {
		const errors: string[] = [];

		// Validate `input`
		if (!("input" in maybeConfig)) {
			errors.push(`${chalk.cyan("input")} is missing`);
		} else if (!maybeConfig.input || typeof maybeConfig.input !== "object") {
			errors.push(
				`${chalk.cyan("input")} should be of type ${chalk.cyan("object")}`,
			);
		} else {
			const input = maybeConfig.input as Record<string, unknown>;

			// Validate `input.directory`
			if (!("directory" in input)) {
				errors.push(`${chalk.cyan("input.directory")} is missing`);
			} else if (typeof input.directory !== "string") {
				errors.push(
					`${chalk.cyan("input.directory")} should be of type ${chalk.cyan(
						"string",
					)}`,
				);
			}

			// Validate `input.glob`
			if (
				"glob" in input &&
				typeof input.glob !== "string" &&
				!(
					Array.isArray(input.glob) &&
					input.glob.length > 0 &&
					input.glob.every((g) => typeof g === "string")
				)
			) {
				errors.push(
					`${chalk.cyan("input.glob")} should be of type ${chalk.cyan(
						"string | string[]",
					)}`,
				);
			}
		}

		if (!("renderers" in maybeConfig)) {
			errors.push(`${chalk.cyan("renderers")} is missing`);
		} else if (!Array.isArray(maybeConfig.renderers)) {
			errors.push(
				`${chalk.cyan("renderers")} should be of type ${chalk.cyan("array")}`,
			);
		} else if (maybeConfig.renderers.length === 0) {
			errors.push(`${chalk.cyan("renderers")} should not be empty`);
		} else if (
			maybeConfig.renderers.some(
				(renderer) => !renderer || typeof renderer !== "object",
			)
		) {
			errors.push(
				`All ${chalk.cyan("renderers[n]")} should be of type ${chalk.cyan(
					"object",
				)}`,
			);
		} else {
			if (maybeConfig.renderers.some((renderer) => !("name" in renderer))) {
				errors.push(`Some ${chalk.cyan("renderers[n].name")} are missing`);
			} else if (
				maybeConfig.renderers.some(
					(renderer) => typeof renderer.name !== "string",
				)
			) {
				errors.push(
					`All ${chalk.cyan(
						"renderers[n].name",
					)} should be of type ${chalk.cyan("string")}`,
				);
			} else if (
				maybeConfig.renderers.some(
					(renderer) => !Object.keys(Renderers).includes(renderer.name),
				)
			) {
				errors.push(
					`All ${chalk.cyan(
						"renderers[n].name",
					)} should be of type ${chalk.cyan(
						Object.keys(Renderers)
							.map((name) => `"${name}"`)
							.join(" | "),
					)}`,
				);
			}

			if (maybeConfig.renderers.some((renderer) => !("options" in renderer))) {
				errors.push(`Some ${chalk.cyan("renderers[n].options")} are missing`);
			} else if (
				maybeConfig.renderers.some(
					(renderer) =>
						!renderer.options || typeof renderer.options !== "object",
				)
			) {
				errors.push(
					`All ${chalk.cyan(
						"renderers[n].options",
					)} should be of type ${chalk.cyan("object")}`,
				);
			} else {
				if (
					maybeConfig.renderers.some(
						(renderer) => !("outputDirectory" in renderer.options),
					)
				) {
					errors.push(
						`Some ${chalk.cyan(
							"renderers[n].options.outputDirectory",
						)} are missing`,
					);
				} else if (
					maybeConfig.renderers.some(
						(renderer) => typeof renderer.options.outputDirectory !== "string",
					)
				) {
					errors.push(
						`All ${chalk.cyan(
							"renderers[n].options.outputDirectory",
						)} should be of type ${chalk.cyan("string")}`,
					);
				}
			}
		}

		if (errors.length) {
			throw new Error(
				`Invalid ${chalk.cyan(this.name)} at ${chalk.cyan(
					this.cwd,
				)}:\n\n- ${errors.join("\n- ")}`,
			);
		}
	}
}
