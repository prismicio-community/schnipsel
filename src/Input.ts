import { join } from "node:path";
import { readFile } from "node:fs/promises";

import chalk from "chalk";
import fm from "front-matter";
import { globby } from "globby";

import { debug, logger } from "./lib";
import { InputConfig, SnippetObject } from "./types";

export class Input {
	private _options: InputConfig;
	private _cwd: string;

	constructor(options: InputConfig, cwd = process.cwd()) {
		this._options = options;
		this._cwd = cwd;
	}

	private async _readSingle(path: string): Promise<SnippetObject | undefined> {
		const filePath = join(this._cwd, this._options.directory, path);
		const raw = await readFile(filePath, "utf8");

		const { attributes, body } = fm<Record<string, unknown>>(raw);

		const errors: string[] = [];

		["name", "description", "prefix"].forEach((key) => {
			if (!(key in attributes)) {
				errors.push(`${chalk.cyan(key)} is missing`);
			} else if (typeof attributes[key] !== "string") {
				errors.push(
					`${chalk.cyan(key)} should be of type ${chalk.cyan("string")}`,
				);
			}
		});
		if (!("scopes" in attributes)) {
			errors.push(`${chalk.cyan("scopes")} is missing`);
		} else if (!Array.isArray(attributes.scopes)) {
			errors.push(
				`${chalk.cyan("scopes")} should be of type ${chalk.cyan("string[]")}`,
			);
		}

		if (errors.length) {
			logger.warn(
				`Invalid snippet file ${chalk.cyan(
					path,
				)} (snippet will be ignored):\n\n- ${errors.join("\n- ")}`,
			);

			return;
		}

		return {
			name: attributes.name as string,
			description: attributes.description as string,
			scopes: attributes.scopes as string[],
			prefix: attributes.prefix as string,
			// Only keep code block content
			body: body.replace(/^.*?```\w*\n+/m, "").replace(/\n+```.*\n?$/m, ""),
		};
	}

	async read(): Promise<SnippetObject[]> {
		debug("Looking for snippet files...");
		const files = await globby(this._options.glob || "**/*.md", {
			cwd: join(this._cwd, this._options.directory),
		});
		debug("%o snippet files found!", files.length);

		debug("Reading snippet files...");
		const snippets = (
			await Promise.all(files.map((file) => this._readSingle(file)))
		).filter(Boolean) as SnippetObject[];
		debug("%o snippet files read!", snippets.length);

		return snippets;
	}
}
