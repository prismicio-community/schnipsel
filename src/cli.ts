import cac from "cac";
import chalk from "chalk";
import exit from "exit";
import latestVersion from "latest-version";
import semver from "semver";

import * as commands from "./commands";
import { NAME, PACKAGE, VERSION, CONFIG_FILE } from "./const";
import { lineBreak, logger } from "./lib";

const cli = cac(NAME);

cli
	.command(
		"",
		`Render Schnipsel project according to current working directory ${chalk.cyan(
			CONFIG_FILE,
		)}`,
		{ allowUnknownOptions: false },
	)
	.option("-w, --watch", "Render in watch mode")
	.action(async (options) => {
		await commands._default(cli, options);
	});

cli
	.command("init", `Init a ${chalk.cyan(CONFIG_FILE)} file`, {
		allowUnknownOptions: false,
	})
	.action(async (options) => {
		await commands.init(cli, options);
	});

cli.version(VERSION);
cli.help(commands.help);

const run = async (): Promise<void> => {
	try {
		cli.parse(process.argv, { run: false });
		await cli.runMatchedCommand();
	} catch (error) {
		if (error instanceof Error) {
			logger.error(error.message);
		} else {
			logger.fatal("");
			console.log(error);
			lineBreak();
		}
		exit(1);
	}

	try {
		const version = await latestVersion(PACKAGE);
		if (semver.gt(version, VERSION)) {
			lineBreak();
			logger.info(
				`A new version of ${chalk.cyan(
					PACKAGE,
				)} is available!\n\n  Install it with:\n  $ npm install --global ${PACKAGE}@${version}\n`,
			);
		}
	} catch (error) {
		// Ignore
	}
};

process.on("unhandledRejection", (error) => {
	logger.fatal("");
	console.log(error);
	lineBreak();
	exit(2);
});

run();
