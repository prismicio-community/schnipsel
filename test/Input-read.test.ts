import { test, expect, vi } from "vitest";

import chalk from "chalk";

import { logger } from "../src/lib";
import { Input } from "../src/Input";

chalk.level = 0;

vi.mock("../src/lib", async () => {
	const {
		logger,
		...rest
	}: Record<string, Record<string, unknown>> = await vi.importActual(
		"../src/lib",
	);

	return { logger: { ...logger, warn: vi.fn() }, ...rest };
});

test("reads input directory and warns about invalid files", async () => {
	const input = new Input({ directory: "__fixtures__/valid/src" }, __dirname);

	expect(await input.read()).toMatchSnapshot();
	expect(vi.mocked(logger.warn)).toHaveBeenCalledTimes(4);
	// @ts-expect-error - Apparently Vitest is messing types(?)
	expect(vi.mocked(logger.warn).calls).toMatchSnapshot();

	vi.mocked(logger.warn).mockRestore();
});
