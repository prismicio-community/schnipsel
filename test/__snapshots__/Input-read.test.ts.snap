// Vitest Snapshot v1

exports[`reads input directory and warns about invalid files 1`] = `
[
  {
    "body": "export const hello = (str: string): void => {
	console.info(\\"Hello %s\\", str);
};",
    "description": "Says hello to the world",
    "name": "Hello World",
    "prefix": "helloWorld",
    "scopes": [
      "typescript",
    ],
  },
  {
    "body": "",
    "description": "Says hello to the world missing",
    "name": "Hello World Missing",
    "prefix": "helloWorldMissing",
    "scopes": [
      "typescript",
    ],
  },
]
`;

exports[`reads input directory and warns about invalid files 2`] = `
[
  [
    "Invalid snippet file emptyFrontMatter.md (snippet will be ignored):

- name is missing
- description is missing
- prefix is missing
- scopes is missing",
  ],
  [
    "Invalid snippet file invalidFrontMatter.md (snippet will be ignored):

- prefix should be of type string
- scopes should be of type string[]",
  ],
  [
    "Invalid snippet file missingFrontMatter.md (snippet will be ignored):

- name is missing
- description is missing
- prefix is missing
- scopes is missing",
  ],
  [
    "Invalid snippet file partialFrontMatter.md (snippet will be ignored):

- prefix is missing
- scopes is missing",
  ],
]
`;
